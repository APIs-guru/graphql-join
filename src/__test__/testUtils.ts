declare var expect;

import {
  Source,
  DocumentNode,
  GraphQLError,
  GraphQLSchema,
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  ExecutionArgs,
  ExecutionResult,

  graphql,
  parse,
  validate,
  execute,
  print,
  buildSchema,
  printSchema,
  formatError,
  specifiedRules,
} from 'graphql';
import * as _ from 'lodash';

import { GraphQLJoinSchema, ProxyContext } from '../index';
import { stubSchema } from '../utils';

expect.addSnapshotSerializer({
  print(val) {
    return typeof val === 'string' ?
      val.trim() :
      JSON.stringify(val, null, 2);
  },
  test(val) {
    return true;
  },
});

function fakeFieldResolver(schemaName: string): GraphQLFieldResolver<any,any> {
  return (
    _1, args: object, _3,
    {fieldName, parentType}: GraphQLResolveInfo
  ) => {
    let result = `${schemaName}::${parentType.name}::${fieldName}`;
    if (Object.keys(args).length !== 0) {
      result += `;args=${JSON.stringify(args)}`
    }
    return result;
  }
}

function makeProxy(schemaName: string, schema: GraphQLSchema) {
  return (queryAST: DocumentNode, variableValues?: object) => {
    const query = print(queryAST);
    expect(query).toMatchSnapshot();
    // FIXME: Should use executeQuery but blocked by
    // https://github.com/facebook/jest/issues/3917
    return graphql({
      schema,
      source: new Source(query, 'Send to ' + schemaName),
      variableValues,
      fieldResolver: fakeFieldResolver(schemaName),
    });
  };
}

type TestSchema = string | { idl: string, prefix?: string };
type TestSchemasMap = { [name: string]: TestSchema };
export function testJoin(testSchemas: TestSchemasMap, joinIDL: string) {
  const remoteSchemas = _.mapValues(testSchemas, (schemaSource, name) => {
    if (typeof schemaSource === 'string') {
      schemaSource = { idl: schemaSource };
    }
    const { prefix, idl } = schemaSource;
    const schema = buildSchema(new Source(idl, name));
    stubSchema(schema);
    return { schema , prefix};
  });

  const joinSchema = new GraphQLJoinSchema(
    new Source(joinIDL, 'Join SDL'),
    remoteSchemas
  );
  const schema = joinSchema.schema;

  expect(schema).toBeInstanceOf(GraphQLSchema);
  expect(printSchema(schema)).toMatchSnapshot();
  return async (
    query: string | { query: string, variableValues: object },
    results?: { [schemaName: string]: ExecutionResult }
  ) => {
    const proxyFns = _.mapValues(remoteSchemas, ({schema}, name) => {
      if (results && results[name]) {
        return (queryAST) => {
          const query = print(queryAST);
          expect(query).toMatchSnapshot();
          return Promise.resolve(results[name])
        };
      }
      return makeProxy(name, schema)
    });

    const queryObj = typeof query === 'string' ?
      { query: query, variableValues: {} } : query;
    const result = await executeQuery({
      schema,
      document: parse(new Source(queryObj.query, 'ClientQuery')),
      variableValues: queryObj.variableValues,
      contextValue: new ProxyContext(joinSchema, proxyFns),
    });

    expect([
      query,
      resultToJSON(result),
    ]).toMatchSnapshot();
  };
}

function executeQuery(args: ExecutionArgs) {
  const validationErrors = validate(args.schema, args.document, specifiedRules);
  expect(validationErrors.join('\n')).toBe('');
  return execute(args);
}

function resultToJSON(result) {
  return {
    ...result,
    errors: (result.errors || []).map(formatError),
  };
}

