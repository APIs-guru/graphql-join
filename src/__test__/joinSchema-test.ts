import { testJoin } from './testUtils';

describe('joinSchema', () => {

  test('Proxy only types referenced inside join schema', async () => {
    const execute = testJoin(
      {
        test: `
          type TypeToIgnore { stub: String, stubSubtype: SubTypeToIgnore }
          type SubTypeToIgnore { baz: String }
          type TypeToProxy { foo: String, bar: SubTypeToProxy }
          type SubTypeToProxy { baz: String }
          type Query {
            typeToIgnore: TypeToIgnore
            typeToProxy: TypeToProxy
          }
        `,
      }, `
        type Query {
          typeToProxy: TypeToProxy @resolveWith(query: "typeToProxy")
        }
        query typeToProxy @send(to: "test") {
          typeToProxy {
            ...CLIENT_SELECTION
          }
        }
      `);

    await execute(`
      {
        typeToProxy {
          foo
          bar {
            baz
          }
        }
      }
    `);
  });

  test('Prefix remote types', async () => {
    const execute = testJoin(
      {
        test: {
          idl: `
            type Query { foo: BarOrBaz }
            union BarOrBaz = Bar | Baz
            type Bar { bar: String }
            type Baz { baz: String }
          `,
          prefix: 'PrEfIx_'
        },
      },
      'schema { query: PrEfIx_Query }'
    );

    await execute(
      `{
        foo {
          ... on PrEfIx_Bar { bar }
          ... on PrEfIx_Baz { baz }
        }
      }`,
      { test: { data: {
        foo: { ___t_test: 'Baz', baz: 'test::Baz::baz' }
      }}}
    );
  });

  test('Merge duplicate types', async () => {
    const execute = testJoin(
      {
        test1: `
          type Foo { foo: String }
          type Query { foo: Foo }
        `,
        test2: `
          type Foo { foo: String }
          type Query { foo: Foo }
        `,
      }, `
        type Query {
          proxyFoo1: Foo @resolveWith(query: "foo1")
          proxyFoo2: Foo @resolveWith(query: "foo2")
        }
        query foo1 @send(to: "test1") {
          foo {
            ...CLIENT_SELECTION
          }
        }
        query foo2 @send(to: "test2") {
          foo {
            ...CLIENT_SELECTION
          }
        }
      `
    );

    await execute(`
      {
        proxyFoo1 { foo }
        proxyFoo2 { foo }
      }
    `);
  });

  test('Merge duplicate abstract types', async () => {
    const idl = `
      type Query { foo: BarOrBaz }
      union BarOrBaz = Bar | Baz
      type Bar { bar: String }
      type Baz { baz: String }
    `;
    const execute = testJoin(
      {
        test1: idl,
        test2: idl,
      }, `
        type Query {
          proxyFoo1: BarOrBaz @resolveWith(query: "foo1")
          proxyFoo2: BarOrBaz @resolveWith(query: "foo2")
        }
        query foo1 @send(to: "test1") {
          foo {
            ...CLIENT_SELECTION
          }
        }
        query foo2 @send(to: "test2") {
          foo {
            ...CLIENT_SELECTION
          }
        }
      `
    );

    await execute(
      `
        query {
          proxyFoo1 { ... on Bar { bar } }
          proxyFoo2 { ... on Baz { baz } }
        }
      `,
      {
        test1: { data: {
          foo: { ___t_test1: 'Bar', bar: 'test1::Bar::bar' }
        }},
        test2: { data: {
          foo: { ___t_test2: 'Baz', baz: 'test2::Baz::baz' }
        }},
      }
    );
  });

});
