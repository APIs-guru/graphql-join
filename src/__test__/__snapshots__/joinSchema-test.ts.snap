// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`joinSchema Merge duplicate abstract types 1`] = `
type Bar {
  bar: String
}

union BarOrBaz = Bar | Baz

type Baz {
  baz: String
}

type Query {
  proxyFoo1: BarOrBaz
  proxyFoo2: BarOrBaz
}
`;

exports[`joinSchema Merge duplicate abstract types 2`] = `
{
  foo {
    ... on Bar {
      bar
    }
    ___t_test1: __typename
  }
}
`;

exports[`joinSchema Merge duplicate abstract types 3`] = `
{
  foo {
    ... on Baz {
      baz
    }
    ___t_test2: __typename
  }
}
`;

exports[`joinSchema Merge duplicate abstract types 4`] = `
[
  "\\n        query {\\n          proxyFoo1 { ... on Bar { bar } }\\n          proxyFoo2 { ... on Baz { baz } }\\n        }\\n      ",
  {
    "data": {
      "proxyFoo1": {
        "bar": "test1::Bar::bar"
      },
      "proxyFoo2": {
        "baz": "test2::Baz::baz"
      }
    },
    "errors": []
  }
]
`;

exports[`joinSchema Merge duplicate types 1`] = `
type Foo {
  foo: String
}

type Query {
  proxyFoo1: Foo
  proxyFoo2: Foo
}
`;

exports[`joinSchema Merge duplicate types 2`] = `
{
  foo {
    foo
  }
}
`;

exports[`joinSchema Merge duplicate types 3`] = `
{
  foo {
    foo
  }
}
`;

exports[`joinSchema Merge duplicate types 4`] = `
[
  "\\n      {\\n        proxyFoo1 { foo }\\n        proxyFoo2 { foo }\\n      }\\n    ",
  {
    "data": {
      "proxyFoo1": {
        "foo": "test1::Foo::foo"
      },
      "proxyFoo2": {
        "foo": "test2::Foo::foo"
      }
    },
    "errors": []
  }
]
`;

exports[`joinSchema Prefix remote types 1`] = `
schema {
  query: PrEfIx_Query
}

type PrEfIx_Bar {
  bar: String
}

union PrEfIx_BarOrBaz = PrEfIx_Bar | PrEfIx_Baz

type PrEfIx_Baz {
  baz: String
}

type PrEfIx_Query {
  foo: PrEfIx_BarOrBaz
}
`;

exports[`joinSchema Prefix remote types 2`] = `
{
  foo {
    ... on Bar {
      bar
    }
    ... on Baz {
      baz
    }
    ___t_test: __typename
  }
}
`;

exports[`joinSchema Prefix remote types 3`] = `
[
  "{\\n        foo {\\n          ... on PrEfIx_Bar { bar }\\n          ... on PrEfIx_Baz { baz }\\n        }\\n      }",
  {
    "data": {
      "foo": {
        "baz": "test::Baz::baz"
      }
    },
    "errors": []
  }
]
`;

exports[`joinSchema Proxy only types referenced inside join schema 1`] = `
type Query {
  typeToProxy: TypeToProxy
}

type SubTypeToProxy {
  baz: String
}

type TypeToProxy {
  foo: String
  bar: SubTypeToProxy
}
`;

exports[`joinSchema Proxy only types referenced inside join schema 2`] = `
{
  typeToProxy {
    foo
    bar {
      baz
    }
  }
}
`;

exports[`joinSchema Proxy only types referenced inside join schema 3`] = `
[
  "\\n      {\\n        typeToProxy {\\n          foo\\n          bar {\\n            baz\\n          }\\n        }\\n      }\\n    ",
  {
    "data": {
      "typeToProxy": {
        "foo": "test::TypeToProxy::foo",
        "bar": {
          "baz": "test::SubTypeToProxy::baz"
        }
      }
    },
    "errors": []
  }
]
`;
